lib/linters.py

```diff
@@ -0,0 +1,627 @@
+    # build a local copy of the linter specific config
+    cn = self.CONFIG_NAME or self.__class__.__name__
+    local_config = deepcopy(self.DEFAULTS)
+    local_config.update(self.global_config.get(cn, {}))
+    self.config = local_config
+    self.setup()
+
+  def setup(self):
+    """setup() may optionally be implemented by each linter"""
+    pass
+
+  def add(self, *args, **kwargs):
+    if 'category' not in kwargs and self.CATEGORY:
+      kwargs['category'] = self.CATEGORY
+    self.errors.add(*args, **kwargs)
+
+  def lint_naming(self, definitions):
+    """Execute the lint check_* methods against the specified definitions."""
+    if not self.config.get('enabled', self.DEFAULT_ENABLED):
+      return
+
+    for itemunit in definitions.networks.values():
+      self.check_network(itemunit)
+
+    for itemunit in definitions.services.values():
+      self.check_service(itemunit)
+
+    self.check_networks(definitions.networks)
+    self.check_services(definitions.services)
+
+  def lint_policy(self, policy):
+    """Execute the lint check_term method against the specified policy"""
+    if not self.config.get('enabled', self.DEFAULT_ENABLED):
+      return
+
+    for header, terms in policy.filters:
+      if not terms:
+        # no terms included -probably just include statements
+        continue
+      for term in terms:
+        self.check_term(policy.filename, header, term)
+
+  def check_network(self, network):
+    """May be implemented to check a given network found in a .net file"""
+    pass
+
+  def check_service(self, service):
+    """May be implemented to check a given service found in a .svc file"""
+    pass
+
+  def check_networks(self, networks):
+    """May be implemented to check all of the networks found in a .net file"""
+    pass
+
+  def check_services(self, services):
+    """May be implemented to check all of the services found in a .svc file"""
+    pass
+
+  def check_term(self, policy, header, term):
+    """May be implemented to check the given term under a given header"""
+    pass
+
+
+@register_linter
+class AddressNumEnforce(BaseLintRule):
+  """Warns about terms with more than a few address tokens."""
+
+  def check_term(self, policy, header, term):
+    addr_attrs = (
+      'source_address',
+      'destination_address',
+      'destination_address_exclude',
+      'source_address_exclude',
+    )
+    for attrname in addr_attrs:
+      addrs = getattr(term, attrname)
+      parents = set([addr.parent_token for addr in addrs])
+      if len(parents) > 4:
+        msg = ("term %s has too many objects in" % term.name +
+             " one of the address fields. " +
+             "Please review and consolidate the addresses")
+        self.add(Severity.WARNING, msg, term.name.lineno)
+
+
+@register_linter
+class RegexNameEnforcer(BaseLintRule):
+  """Ensures objects and terms following naming conventions."""
+
+  DEFAULTS = {
+    'NETNAME': r'^([A-Z][A-Z0-9_]+|h_[0-9a-f\.]+|n_[0-9a-f\.]+_[0-9]+)$',
```

Comment:
Wrong hanging indentation here also, add two spaces to put them into the correct location.